http/client/src/main/java/org/wildfly/security/http/client/ElytronHttpClient.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.http.client;
+
+import org.wildfly.security.http.client.exception.ElytronHttpClientException;
+import org.wildfly.security.http.client.mechanism.basic.ElytronHttpClientBasicAuthMechanism;
+import org.wildfly.security.http.client.mechanism.bearer.ElytronHttpClientBearerAuthMechanism;
+import org.wildfly.security.http.client.mechanism.digest.ElytronHttpClientDigestAuthMechanism;
+import org.wildfly.security.http.client.utils.ElytronHttpClientCredentialUtils;
+import org.wildfly.security.http.client.utils.ElytronMessages;
+
+import javax.net.ssl.SSLContext;
+
+import static org.wildfly.security.http.HttpConstants.OK;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class ElytronHttpClient {
+
+    private HttpClient httpClient;
+
+    public ElytronHttpClient() {
+        this.httpClient = HttpClient.newHttpClient();
+    }
+
+    /**
+     * Used to connect to the secured uri and return the response based on that.
+     */
+    public HttpResponse connect(HttpRequest httpRequest) throws IOException, InterruptedException, URISyntaxException {
+        addSSLContextToHttpClient(httpRequest.uri());
+
+        HttpResponse response = getResponse(httpRequest);
+        if (response.statusCode() == OK) {
+            return response;
+        }
+
+        String authenticateHeader = getAuthenticateHeader(response);
+        if (authenticateHeader == null) {
+            throw new ElytronHttpClientException(ElytronMessages.log.responseHeaderExtractionFailed());
+        }
+        HttpRequest authRequest = null;
+
+        String challenge = getFirstChallenge(authenticateHeader);
+        String mechanismType = getMechanismType(challenge);
+
+        switch (mechanismType) {
+            case "basic":
+                authRequest = ElytronHttpClientBasicAuthMechanism.evaluateMechanism(httpRequest);
```

Comment:
@keshav-725 Just a minor, we can create an interface for all client HTTP mechanisms that will contain an evaluateMechanism method